### DISCUSSION TITLE
Support Planetscale

### DISCUSSION THREAD

**Discussion by Syed-Sheharyar** [directus/directus#19419] [Feature Requests]
### Summary

Firstly I am unsure whether Planetscale is supported or not. Because on this page https://directus.io/engine
![Web capture_13-8-2023_10343_directus io](https://github.com/directus/directus/assets/96941588/30b64cd8-e0da-48be-9843-ca66b34035c9)
You see that Planetscale is also listed in the supported databases with its logo. I can't find any single example of using Planetscale in your docs. **Also, seems like you only mentioned Planetscale but not supporting it which is misleading**. In this discussion, https://github.com/planetscale/discussion/discussions/274 everybody says that Directus does not support Planetscale. Also, the documentation should have examples of managing migrations when used with popular ORMs like Prisma and Drizzle. Because in big projects developers need control of their underlying database not everything can be managed through Directus. Your product is great but documentation sucks sometimes. There is also a discussion present on your subreddit [Planetscale support ](https://www.reddit.com/r/Directus/comments/14yu06i/directus_planetscale/)
I am not the only one who needs Planetscale support. **Now, your upvote is required to make this discussion reach more than 15 upvotes. In the end, I want this message from the maintainer.**

> Heya! Thanks for opening this feature request!
> 
> This feature request has received over 15 votes from the community. This means we'll move this feature request to the Under Review state!
> 
> The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.
> 
> For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html).

### Basic Example

Change in Migrations and defining SQL schema.
`CREATE TABLE users (
    id INT PRIMARY KEY AUTO_INCREMENT,
    name VARCHAR(80) NOT NULL,
    email VARCHAR(80) NOT NULL,
    password VARCHAR(255) NOT NULL
);`

## Normal way:
`CREATE TABLE posts (
  id INT PRIMARY KEY AUTO_INCREMENT,
  author_id INT,
  title VARCHAR(255) NOT NULL,
  content TEXT NOT NULL,
  FOREIGN KEY (author_id) REFERENCES users(id)
);`

## Planetscale way:
`CREATE TABLE posts (
  id INT PRIMARY KEY AUTO_INCREMENT,
  author_id INT,
  title VARCHAR(255) NOT NULL,
  content TEXT NOT NULL,
  KEY (author_id)
);`

### Motivation

I am going to use Planetscale as my database in my next big project and want to integrate it with Directus. The only thing that motivates me is their new scaler pro plan. https://planetscale.com/pricing. No reads write limitations. And, yeah it has so many features that I can't even list here like scalability, automatic backups, Planetscale Boost, and more. Due to their generous free tier even SaaS and startup businesses used it as a database of their choice.

### Detailed Design

The only thing you need to do is to change the relationship mechanism. Planetscale doesn't support foreign key constraints, but it does support relationships. You need to implement relations in Planetscale style. Using Planetscale KEY you found docs on their site.

### Must Have:
Support for Planetscale relations

### Should Have:
Don't allow non-existent data to be present in the relational table at least from the data studio and from everywhere we fill the rows.

**For example,** _If I had two tables one "users", second "posts" and two user's user 1, user 2. In this case, only users 1 and 2 should be referenced in the Posts table as foreign keys but Planetscale does not support foreign key constraints. I even reference user 3 in one of my posts in the Posts table which foreign key constraints are not allowed. If user 3 does not exist in the users table I should not reference it in the Posts table. Before adding data from the data studio to the collection make sure that any value added to the foreign key column should exist in the Primary key table through which the relation is created._
Remember, Planetscale doesn't support foreign key constraints but supports foreign keys called KEYS in their term.

### Could Have:
Manage deletion of orphan rows as there are no foreign key constraints so we can't use **ON CASCADE DELETE**. For example, if the server has low traffic or low pressure then automatically deletes some orphan rows. Also, give the manual one-click option to delete specified orphan rows 100, 200, or if not so many so, delete all at one click.

### Won't Have:
Won't make subsequent queries to the database and cache data because the pricing of Planetscale is based on read writes. If we navigate from one page to another don't make a query for the same data.

### Drawbacks

Implementing it is not too easy but not hard as well. It is not going to be a breaking change. You are using knex.js under the hood which doesn't support Planetscale. But I found this GitHub repo https://github.com/hfaulds/knex-planetscale. It might help you. I don't know how everything works internally and just suggesting some ideas and want to keep discussing it until it resolves. It should not be too hard it is the same as MySQL without foreign key constraints. **[ Vitess ]**

### Alternatives

No Alternatives, impact is that you lost a big userbase of Planetscale that might contribute back to you or sponsor you.

### Adoption Strategy

It can't affect your existing projects. In the end, you are supporting one more database resulting in the growth of the Directus userbase. You need to write migrations in a way that Planetscale supports at the end then we use it with any ORM like Drizzle or Prisma because both support Planetscale by taking care of migrations. if any official article is present on your website as a guide.

### Unresolved Questions

If you are not supporting Planetscale then why misleading people by showing Planetscale in supported databases?

**Comment by rijkvanzanten** [comment]
> You see that Planetscale is also listed in the supported databases with its logo

That graphic is to illustrate how Directus is architected and works, but not a conclusive list of everything that's supported. There's a couple of vendors missing as well. 

The reason Planetscale ended up on that list is because the graphic was made when Planetscale was announced, but not yet released. At the time, MySQL compatibility was communicated, which we understood as.. MySQL compatibility which in turn would've meant Directus would just work with it as well 😅 

We're currently re-architecting how data is fetched from sources altogether; support for Planetscale is in the works as well. That being said, it's definitely not as simple as _The only thing you need to do is to change the relationship mechanism._ 😉

**Reply by Syed-Sheharyar** [reply]
Thanks for replying. 
> The only thing you need to do is to change the relationship mechanism.

Looks easy on paper but is hard to implement. It means I need to wait 4 to 6 months or more for Planetscale. 😄

**Comment by rollemira** [comment]
This would be really cool. But I doubt this feature could exist without community support.

**Comment by cesarsalesgomes** [comment]
Planetscale recently added support for foreign key constraints (https://planetscale.com/blog/announcing-foreign-key-constraints-support), which was the first barrier when bootstrapping Directus with Planetscale, due to an error that Directus throws because it needs the constraint.

The error no longer occurs, now going to the one shown in the image:

![Planetscale](https://github.com/directus/directus/assets/6998275/c8070647-ec00-42f4-abe8-c3e274d070a1)

**Comment by cesarsalesgomes** [comment]
Directus recent 10.9.0 update apparently fixed the first previous error, and thus it was possible to finish the installation 😌, but with some manual adjustments:

1. Activate 'Foreign key constraints' beta feature on the Planetscale database.
2. Bootstrap the Directus project with the 'NPM Installation' on the [Quickstart guide](https://docs.directus.io/getting-started/quickstart.html) using the Planetscale credentials.
3. A SSL error will occur, as mentioned on this [issue](https://github.com/directus/directus/discussions/13536). 
4. Finish the script and create a `.env` file in the created directory with the necessary [database env vars](https://docs.directus.io/self-hosted/config-options.html#database) and the one that ignores the SSL certificate (`DB_SSL__REJECT_UNAUTHORIZED="false"`)
5. Bootstrap the project with `npx directus bootstrap` to initiate the migrations
6. For some reason, migrations that create columns, and change them in the same file, are throwing errors. Going to the migration file that the error occurs at `node_modules\@directus\api\dist\database\migrations` and just commenting on the knex column creation script (as it has already been created) will fix the migration when running again. Around 4 migrations will occur this same error.
7. After all migrations finish successfully, start Directus with `npx directus start` 😃

In this process, the admin user is not created, as an error occurs on the CLI installation. Manually inserting it into the base copying the password of an application that was started correctly, replicating the same KEY and SECRET env vars, is a workaround.

I believe that by finding a way to install avoiding the SSL error (for administrator to be created), and resolving the migration errors in a new version, it will be possible to bootstrap Directus with Planetscale perfectly 🚀

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
